design/providers/kubernetes_secrets_storage_csi/conjur_csi_provider_solution_design.md

```diff
@@ -0,0 +1,340 @@
+
+The `CSIDriverProviderServer` interface includes the `Mount` and `Version` methods:
+
+1. `Mount`: Invoked by the Secrets Store CSI Driver during the volume mount phase. The `MountRequest` object passed to this method contains several important fields:
+
+   - `Attributes`: A string containing JSON-encoded attributes. These attributes contain properties needed for the provider to fetch secrets, such as the `authnUrl` and secret IDs. This is a flexible mechanism that allows for a variable set of attributes based on the provider's requirements.
+
+   - `Secrets`: A string containing JSON-encoded secrets. These secrets contain sensitive data such as the Kubernetes service account token, which is used for authenticating with Conjur via the `authn-jwt` authenticator.
+
+   - `TargetPath`: The path where the Secrets Store CSI Driver expects the secrets files to be written. The provider is responsible for fetching the requested secrets and writing them to this path.
+
+   - `Permission`: A string containing JSON-encoded file permissions. These are the permissions that should be applied to the files containing the secrets.
+
+   The `Mount` method's responsibilities are to authenticate with the Conjur instance, fetch the secrets specified in the `Attributes`, write these secrets to files at the `TargetPath`, and then return a `MountResponse` to the Secrets Store CSI Driver. The `MountResponse` includes metadata about the fetched secrets, such as their IDs and versions.
+
+2. `Version`: Invoked by the Secrets Store CSI Driver to discover the provider's version information. This method returns a `VersionResponse` object that includes the version of the provider protocol, and the name and version of the provider's runtime.
+
+By implementing this gRPC server and the `CSIDriverProviderServer` interface, we can create a Conjur provider for the Kubernetes Secrets Store CSI Driver. This provider will handle fetching secrets from the Conjur instance and passing them back to the Secrets Store CSI Driver, which in turn makes them available to Kubernetes workloads.
+
+Helm charts will be created for deploying the CyberArk Conjur provider. This provides a streamlined, automated, and repeatable deployment process.
+
+### UX
+
+The user interaction with our solution will mainly be through Kubernetes manifests and Helm.
+
+#### SecretProviderClass
+
+`SecretProviderClass` is a Kubernetes custom resource that stores the secret provider configuration. The user will define a Kubernetes `SecretProviderClass` object that refers to the CyberArk Conjur provider and specifies necessary parameters for it. This would look something like the following:
+
+```yaml
+apiVersion: secrets-store.csi.x-k8s.io/v1alpha1
+kind: SecretProviderClass
+metadata:
+  name: db-credentials
+spec:
+  provider: conjur
+  parameters:
+    authnUrl: https://conjur-auth.some-org.com
+    applianceUrl: https://conjur.some-org.com
+    account: some-org
+    authnLogin: path/to/some-policy/some-host
```

Comment:
I think we should align the parameter names to what we have in Secrets Provider to reduce confusion.
I see that there we call it `authnIdentity` ([reference](https://docs.conjur.org/Latest/en/Content/Integrations/k8s-ocp/cjr-k8s-secrets-provider-ref.htm?tocpath=Integrations%7COpenShift%252FKubernetes%7CApp%20owner%253A%20Set%20up%20workloads%20in%20Kubernetes%7CSet%20up%20workloads%20(cert-based%20authn)%7CSecrets%20Provider%20for%20Kubernetes%7C_____3#PushtoFileannotationsKubernetesSecretsenvVars))